PEP 788: Add some minor clarifications (#4391)
python/peps: peps/pep-0788.rst

    proceed. The API was designed as "it'll block and only return once it has
    the GIL" without any other option.

For this reason, we can't make any real changes to how :c:func:`PyGILState_Ensure`

> works for finalization, because it would break code. Similarly, threads

Reviewer: ```suggestion
works for finalization, because it would break existing code. Similarly, threads
```

or in between bytecode instructions, based on :func:`sys.setswitchinterval`.

A new, internal field will be added to the ``PyThreadState`` structure that
determines if the thread is daemon. Before finalization, the interpreter
will wait until all non-daemon threads call :c:func:`PyThreadState_Delete`.

For backwards compatibility, all thread states created by existing APIs,

> including :c:func:`PyGILState_Ensure`, will remain daemon by default.

Reviewer 2: Perhaps include the rationale you posted on Discourse here?

Author: Hm, I thought the [section](https://peps.python.org/pep-0788/#we-can-t-change-finalization-behavior-for-pygilstate-ensure) on this was pretty clear. I'll just link to that.

Reviewer 2: I think the confusion is that "finalisation behaviour" section makes no reference to daemon status, so it's not immediately obvious that this is the reason. The first explicit mention of backwards compatability is this section in *Specification*, so it might be useful to add some prose in earlier sections, to give the reader more context.

Author: I linked to it there, and also added some additional information to the section in the motivation. Let me know if that's any better.

*****************************

As stated in the Motivation_, there are many cases where it's desirable

> to call Python in an asynchronous callback. In that case, it's not safe to

Reviewer 3: ```suggestion
to call Python in an asynchronous callback. In such cases, it's not safe to
```

You refer to multiple, plural "cases" in the previous sentence, so it seems clearer and less potentially confusing to match that here.

> call :c:func:`PyInterpreterState_Hold`, because it's not guaranteed that
> :c:func:`PyThreadState_Ensure` will ever be called, which would deadlock
> finalization.

Reviewer 3: Its somewhat ambiguous from the text if finalization becomes deadlocked if `PyThreadState_Ensure` _is_ called, or if it is not called. Assuming its the latter, I'd write:

```suggestion
call :c:func:`PyInterpreterState_Hold`, because it's not guaranteed that
:c:func:`PyThreadState_Ensure` will ever be called.
If not, finalization becomes deadlocked.
```
